File: Questions/653. Two Sum IV/main.cpp
#include <unordered_set>
class TreeNode {
    public: 
        int val;
        TreeNode* left;
        TreeNode* right;
        TreeNode(int val) : val(val), left(nullptr), right(nullptr) {}
        TreeNode(int val, TreeNode* left, TreeNode* right) : val(val), left(left), right(right) {}
};

class Solution {
    public:
        bool dfs(TreeNode* root, int k, std::unordered_set<int>& s) {
            if(!root) return false;
            if (s.count(k - root->val)) return true;
            s.insert(root->val);
            return dfs(root->left, k, s) || dfs(root->right, k, s);
        }

        bool findTarget(TreeNode* root, int k) {
            std::unordered_set<int> s;
            return dfs(root, k, s);
        }
};

int main() {
    TreeNode* node1 = new TreeNode(0);
    return 0;
}

File: Questions/653. Two Sum IV/main.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
def find_target(root: TreeNode | None, k: int) -> bool:
    seen = set()
    if not root:
        return 

    def dfs(node: TreeNode | None) -> bool:
        if not node:
            return False

        if k - node.val in seen:
            return True
        
        seen.add(node.val)

        return dfs(node.left) or dfs(node.right)
    
    return dfs(root, k)

